cEP-0032: Enhance Websites Architecture (#186)
coala/cEPs: cEP-0032.md

      organization, therefore an addition of feedback form to community website
      will be helpful.

   *Following Validation checks will be performed while:*
      1. **Participate as a mentor**
         * A Developer or Maintainer (​ not Newcomer​ )
         * GSoC Cleared (in any year)
         * GCI Cleared (in any year) (optional)
      2. **Participate as a GSoC participant**
         * cEP Merged
         * Project Idea PR merged
         * At-least one accepting Review ​from any of the maintainers
         * Issue that gives idea to solve project proposal has a label of
         “Project/Proposal”
         * Valid project link to projects website 
      3. Only **coala-developers** (i.e. members part of coala developers team)
      will be able to add new google forms and add/update calendar events.
      4. **Joining the coala community** (*newcomers*)
         * Hasn't joined GitHub recently (i.e. not created account recently)
         * verify their email address using an email-verifier-api
         * has atleast 1 source repositories on GitHub
         * has done any of the github training exercises etc.

> 3. **Login/Signup using GitHub**

Reviewer: @jayvdb, needed your views on if this is required or needed to be done?

Reviewer: Please also have a look at [this pr discussions](https://github.com/coala/projects/pull/719) when you get time, we needed your suggestions there too.

Reviewer 2: This would be good.  @Jazzband has this for their website.
The problem will be in getting it hosted, if that is required, so definitely plan on it being an optional feature, and ensure it can be tested on your own fork with a Heroku-type hosting, because we will not deploy it until it is finished and we have time to set up an org instance of the backend.

Author: Yeah! we will get problem in getting it hosted because the authentication would work only if its a dynamic website which is currently being served as a static website to users

Author: @jayvdb Tried to host community website on [heroku](https://kvgarg-community.herokuapp.com/) but unfortunately get a `KeyError`. It's running completely fine when running heroku locally but on deploying to server its raising an exception

Reviewer 3: @KVGarg On what file or line are you getting the exception? Could you show the full error message the python interpreter gives you?

Author: `python-interpreter` isn't giving any error because git remotes are configured locally. But they aren't being configured on heroku.
Heroku deployed project link - https://kvgarg-community.herokuapp.com/

Reviewer 3: Based on the debug info, this is what I see:
```
When `name` is None, fetch current repo of current org,
    otherwise, fetch a specified repo of current org.
    """
    # Use environment variable URL when Netlify detected
    if not name and os.environ.get('REPOSITORY_URL'):
        return os.environ.get('URL')
    url = get_remote_url() ...
    owner = url.owner
    path = name if name else url.name
    if hoster is not None:
        deploy_url = 'https://%s.%s.io/%s' % (owner, hoster, path)
    elif url.resource == 'github.com':
```
If the git remotes are not configured on heroku, try calling `git_remote_url(“URL”)` in `git.py` with a URL parameter that points to any other location where the remotes are configured. If there is no location, maybe they need to be configured also.

Reviewer 3: As I can see this is covered for Netlify. Maybe if you change the provider from heroku to Netlify, it can solve the problem immediately.

Author: `netlify` is a static site generator, right? will that allow us to host it? Try to deployed django forms on netlify but they were not getting deployed

> As I can see this is covered for Netlify. Maybe if you change the provider from heroku to Netlify, it can solve the problem immediately.

Will try to do this!
> If the git remotes are not configured on heroku, try calling git_remote_url(“URL”) in git.py with a URL parameter that points to any other location where the remotes are configured. If there is no location, maybe they need to be configured also.

Author: > 
> 
> As I can see this is covered for Netlify. Maybe if you change the provider from heroku to Netlify, it can solve the problem immediately.

I guess deploying on netlify can be achieved by netlify provided features https://www.netlify.com/docs/authentication-providers/ 😮

Reviewer 3: Yes, I haven’t worked with that, but it seems it’s easy.

Author: > As I can see this is covered for Netlify. Maybe if you change the provider from heroku to Netlify, it can solve the problem immediately.

@jayvdb @sks444 @margobra8 
I'm able to resolve this problem by setting two environment variables - REPOSITORY_URL and URL which are kind of necessary for coala Community repo. After adding them, the authentication is working.
Can be checked at: https://kvgarg-community.herokuapp.com/accounts/github/login/ after getting authorized it will return to homepage

Author: > Will look on netlify OAuth too.

Preview: https://5d0b80bdafb942000880f8cd--determined-goldstine-6dc504.netlify.com/auth/

But If we're sticking with static website It will be difficult to maintain user-session because every time user refreshes the web-page or re-directs to some another page the session will be no more exist. And hence no benefit.

Author: This can have benefit if I guess `Session` values can be added while a user gets logged in and use that until user is on the website.

      community website can be treated for safe side. As well as it will help
      contributors by not spamming on coala main gitter channel for getting
      assigned on issue. Let's just keep that channel for coala related talk.

### Improvements

According to ​study published by adobe, 38% of people will stop engaging with a
website if the content/layout is unattractive. Therefore, this section
describes some major improvements in coala community website. Following are the
imporvements:

1. **UI/UX Design improvement**
   1. Add *materialize.css*
   2. Re-design all webpages
   3. Gamification leaderboard
        Changing format from a *div* block for each contributor to a table
        format will give a better look from user point of view.

2. **Update README.md providing more useful information**

      1. Updated Header image

            The new header image will symbolise the meaning of *creating,
            supporting and growing* community.
      2. *Why we need this webiste? (Role of community)*
      3. Describing features of coala community website

3. **Update coala newcomer docs about automation features**

> ## coala [Webservices website](http://webservices.coala.io/)

Reviewer: Have a section in this something like `Fixing the existing webservices website`, So that we get the real data at https://coala.io/#/getinvolved and community website can use the coala webservices instead of using [my raw git repo](https://github.com/coala/community/blob/07d6d512f84b6b6016ce9dd85f5255110b940f89/data/webservices.py#L12).

P.S. Fixing the existing webservices website with auto deployment on new prs merge should be of highest priority. Mention it in the beginning of coding phase one timeline.

Author: > auto deployment on new prs merge

This should already work! I've seen merged prs and the pipeline is being triggered by repo maintainer. I was researching on this before bonding period, all i found is that some bugs were there which gitlab community fixed. It's still not confirmed whether its working or not. Also, there is no error in `.gitlab-ci.yaml` which isn't causing pipelines to get triggered automatically.
I would like to verify whether this issue still exists or not! Because its directly related with either upstream or settings configured for this repository in which @jayvdb can help us

Author: > Fixing the existing webservices website

I will add this section

Author: >  Mention it in the beginning of coding phase one timeline.

Added in coding phase 1 milestone

| Type     | Feature                                               |

## Abstract

This document describes the features that will increase the usability and
easiness for all the open-source enthusiasts by enhancing and improving the
coala Community website as well as coala Projects website for newcomers,
developers as well as for other communities. In addition, new API Endpoints
will be created in coala webservices. These new endpoints will manage the user
-database as well as coala database in a more structured form.

## Introduction

A website plays an essential role for an open-source community because it
spreads the idea of offerings that community provide to other organizations,
with information about the active community by exploring their skills as well
as their precious contributions to the organization. And to provide such
offerings this project aims to provide useful features to community members as
well as to other developers and organizations. They can use the below
mentioned features to be a part of coala community and can even get information
about coala community and its contributors who have been part of SoC programs.

> ## coala [Community website](community.coala.io/)

Reviewer 4: The link is not working :)

### Enhancements

The listed enhancements will ease the navigatibility and usability for
newcomers, developers and to all others who wanted to be a part of coala
organization. It will help community members to know about their performance
in-comparison to other members. Following are the new enhancments that will be
added to coala community website:

1. **Display a webpage for-**
   1. **organization teams**

>         This will show information about all [coala teams](https://github.com/orgs/coala/teams).

Reviewer 4: Sounds interesting, but that link is private and only available to the community members. Not sure if it is a good idea to make it public.

Author: https://github.com/coala/community/issues/12#issuecomment-457676443
Took idea from https://github.com/coala/community/issues/12

Author: the web-page will be showed to just logged-in users, now

        It will definitely not show what permissions are available to team members
        if a developer is a part of that team. But, it will definitely provide
        information like *What does this team aim to do?*, *Members count* etc.
        Clicking on a **team name** will redirect the user to either on GitHub
        showing all team members or a new webpage which will show all the team
        members of this coala team.

   2. **contributors' geolocation in form of a map**

        Showing the *global influence of an organization* around the world of open-
        source organization can be achieved by showing the contributors' geolocation.
        It can be shown in the form a world map which shows the geolocaton of all
        contributors that are members of an organization or have been contributing
        to the organization.

   3. **mentors**

        It will be good if community website can show all the potential mentors that
        organizaton has. It will allow other organization to get to know more about
        them.

> 2. **Add forms for -**

Reviewer: I am skeptic about this section too.

Not sure if we need all these forms, needed @jayvdb confirmation.

Reviewer 2: See @jazzband

3. **Login/Signup using GitHub & GitLab**

      The authentication will be a way to avoid spam activities or to avoid any
      unwanted updation by newcomers in the database. Also, we can’t use ORG
      API Token for updating user related information as there will more tasks
      which have to completed first. So, using user authentication we can get
      user created API Token to update its related information.

      Adding auth will allow us to make sure that newcomer can’t:

      * Upload any google form
      * Participate as a mentor
      * Update the database
      * Apply for promotion to developer for any other user
      * Assign someone else to the issue

4. **Manage community members data**

      We can't use all the time organization access token to manage the
      contributors live data. Using an access token created by user in there
      account, will allow us to manage their data. Using the access token, the
      user contribution to the coala organization can be maintained in our
      database. Everytime a user logs in a cron job will be scheduled to update
      the user data in the database.

5. **Gamification leaderboard on homepage**

>       A leaderboard that shows up top 10(or more) contributors of the curren

Reviewer 4: ```suggestion
      A leaderboard that shows up top 10 (or more) contributors of the current
```

      month (a drop-down having options with the weekly, monthly or yearly view).
      NON-GCI & GCI participants will have a separate leaderboard.

6. **coala contributor profile page**

      Interactive contributor profile page with various features like the
      statistics 2D or 3D graph that will show hike in contributions, activities,

>       overview of contributions and many more. Contributor will provided with

Reviewer 4: ```suggestion
      overview of contributions and many more. Contributors will be provided with
```

      community website can be treated for safe side. As well as it will help
      contributors by not spamming on coala main gitter channel for getting
      assigned on issue. Let's just keep that channel for coala related talk.

### Improvements

According to ​study published by adobe, 38% of people will stop engaging with a
website if the content/layout is unattractive. Therefore, this section
describes some major improvements in coala community website. Following are the
imporvements:

1. **UI/UX Design improvement**
   1. Add *materialize.css*
   2. Re-design all webpages
   3. Gamification leaderboard
        Changing format from a *div* block for each contributor to a table
        format will give a better look from user point of view.

2. **Update README.md providing more useful information**

      1. Updated Header image

            The new header image will symbolise the meaning of *creating,
            supporting and growing* community.
      2. *Why we need this webiste? (Role of community)*
      3. Describing features of coala community website

3. **Update coala newcomer docs about automation features**

> ## coala [Webservices website](http://webservices.coala.io/)

Reviewer 4: ```suggestion
## coala [Webservices website](https://webservices.coala.io/)
```

### Enhancements

1. **New database models with endpoints**

      To enhance the website architecture these new models will play a vital
      role. It will manage all the required database records that will shown on
      coala community and projects website with which the user will be
      interacting. Following new models will be added to coala webservice api-

      * Calendar
      * Google forms
      * GSoC
      * Mentors
      * A table with models name with updated_time
      * Feedback

      In addition to above new models to be created, the existing models will be
      improved to manage more database of the organization and contributor. The
      existing models that will be improved are-

      * Organization Teams
      * Contributor

> 2. **Keep contributor's geolocation cluster map up-to-date**

Reviewer 4: ```suggestion
2. **Keep contributors' geolocation cluster map up-to-date**
```

      To keep up-to-date contributors' geolocation, a new cron job will be added
      to coala webservices api that will refresh the cluster map with updated
      geolocation of users. Then this map will be included on the coala Community
      website to show the global influence if the organization.

3. **Website homepage to display information about new enhancements**

      The information related to new endpoints created to enhance the websites
      architecture like *endpoint name, url and what it does?* will be added to
      coala Webservices homepage.

### Improvements

**Fix existing *website***

The exisiting website is having some flaws which needs to be fixed-

1. Merged PRs are not being auto-deployed on Docker, instead the pipeline is
being triggered manually to upload it on Docker by the maintainer.
2. coala Webservices API is not being used by coala websites to fetch
organization and contributors related data which is being displayed to users on
both the websites. Instead it is getting static data from

> [@sks444 RawGit repository](https://github.com/coala/community/blob/master/data/webservices.py#L12).

Reviewer 4: ```suggestion
[@sks444 RawGit repository](https://github.com/coala/community/blob/5d83d053da31be6585dcfac18b998485c21950fc/data/webservices.py#L12).
```

> ## coala [Projects website](projects.coala.io/)

Reviewer 4: Link is not working :)

      * Show all students that have been selected for GSoC in past years.
      * Design - A Collage containing all GSoCers for a year or
      * just display like admins/mentors are being shown under mentors tab

### Improvements

1. **Re-design homepage**
2. Add **UI/UX design** to GSoC project [reports page](http://projects.coala.io/reports)
3. **Responsive UI** of a GSoC report page (eg. [Project Status Report](http://projects.coala.io/GSoC/2018/StatusReport/NitanshuVashistha.html))
4. **Mark issues state** in project description popup

      The background color of *related issues* mentioned in project description
      popup should be decided on their current state.

## Implementation

### Contributors' Geolocation

* It is important for any organization to show its global influence by showing
how people from different countries come together. 
* They join open source projects and make that project more usable or great so
that it can aid other developers as well as communities.
* To stay updated, the map will be regularly updated using a ***cron job***.
* The following code will be used to create a cluster map to show coala members
geolocation.

>       from github import Github

Reviewer 4: use ```python

### New API Endpoints in coala-webservices

After the models have been created, we can connect websites to new API endpoints
and their backend to add some data which can be further used to display on
websites.

      urlpatterns = [
      path('^forms/', AddGoogleForms.as_view(), name='Add Google Form'),
      path('^feedback/', AddFeedback.as_view(), name='Add Feedback'),
      path('^gsoc/mentors/', AddMentor.as_view(), name='Add Mentors'),
      path('^gsoc/participant', AddGSoCParticipant.as_view(), name='Add new participant'),
      path('^users/<str:USERNAME>', UserProfile.as_view(), name='User profile'),
      ]

Each endpoint will be have its own backend functionality to add data to the
models database. And, for issues related data I will be creating a cron Job.
A sample code snippet for GSoC Participant and Google Forms

* ***GSoC***
  
  * A database model to have information about all past GSoC students.
  * The contributor can add information to database from their profile section
  after or during GSoC.
  * The POST method will be having rudimentary checks to avoid spamming.

  * On API endpoint following view(*in views.py*) will be called:

>             from .serializers import GSoCParticipantSerializer

Reviewer 4: Use ```python

            class Meta:
                  model = Contributor
                  fields = '__all__'

            def validate(self, data):
                  # Validate POST data
                  # Evaluate the google form
                  # 1. Get user latest entry from response spreadsheet
                  # 2. Valid data i.e.
                  #        - username
                  #        - solved newcomer issue
                  #        - solver low difficult issue
                  #        - reviewed two PRs
                  return data

* Validating this data will definitely require some amount of time.
  This can be achieved in two ways:
  1. One way is that user chooses to get promoted. Then, the API call will be
  made in background and user can’t do any work until the validating process is
  completed.
  2. Other way is create a database model which will have usernames with other
  information. And, then a cron Job will be executed for all such usernames to
  validate their information. The job will be executed for only those username
  who job scheduled was pending.

      I would like to go for second one during coding if I get enough time to
      achieve this.

> ### Authentication using GitHub

Reviewer 3: I don't think auth should be done this way. `django-allauth` has other methods of login that support other platforms more suitable than GitHub. Sould look into any other OAuth provider.

Author: Can you help me out by suggesting other OAuth providers that can help us to achieve this task?

Reviewer 3: Maybe if we leave the new endpoints as a pub-sub service rather than requiring auth when there are forms. I think that rather than finding other providers, it is more suitable to focus at one approach. Either go for form-based and then assume if completion is necessary to create coala auth grants or for GitHub. I think the `django-allauth` approach is ok but make sure that both GitHub and GitLab are implemented.

-database as well as coala database in a more structured form.

## Introduction

A website plays an essential role for an open-source community because it
spreads the idea of offerings that community provide to other organizations,
with information about the active community by exploring their skills as well
as their precious contributions to the organization. And to provide such
offerings this project aims to provide useful features to community members as
well as to other developers and organizations. They can use the below
mentioned features to be a part of coala community and can even get information
about coala community and its contributors who have been part of SoC programs.

## coala [Community website](https://community.coala.io/)

### Enhancements

The listed enhancements will ease the navigatibility and usability for
newcomers, developers and to all others who wanted to be a part of coala
organization. It will help community members to know about their performance
in-comparison to other members. Following are the new enhancments that will be
added to coala community website:

1. **Display a webpage for-**
   1. **organization teams**

>         This will show information about all [coala teams](https://github.com/orgs/coala/teams).

Reviewer 3: That link won't work if visited outside of coala, change accordingly. Maybe only display info that is public, such as org members

Author: ~~Yeah! It won't work for a non-member. Will make that change~~
Sorry! I took that suggestion in some wrong direction. `@jayvdb` proposed the idea of having a page for all teams and I too think that atleast newcomers, developers and admins
team must have a page.
